core/ir/aarch64/codec.c

```diff
@@ -1556,6 +1557,113 @@ encode_opnd_cond(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
     return encode_opnd_int(12, 4, false, 0, 0, opnd, enc_out);
 }
 
+/* fpimm8: immediate operand for SIMD fmov */
+
+static inline bool
+decode_opnd_fpimm8(uint enc, int opcode, byte *pc, OUT opnd_t *opnd)
+{
+    /* See Arm Architecture Reference Manual
+     *
+     * Immediate is encoded as 8 bits. Bits 5->9 and 16->18. LSB is bit 5:
+     * imm8 = a:b:c:d:e:f:g:h (LSB)
+     *
+     * Half-precision (v8.2)
+     * --------------
+     *
+     * imm16 = imm8<7>:NOT(imm8<6>):Replicate(imm8<6>,2):imm8<5:0>:Zeros(6);
+     *         a:~b:bb:cdefgh:000000
+     *
+     * datasize = if Q == '1' then 128 else 64;
+     * imm = Replicate(imm16, datasize DIV 16);
+     *     = imm16:imm16:imm16:imm16                         (Q=0 -> 64)
+     *     = imm16:imm16:imm16:imm16:imm16:imm16:imm16:imm16 (Q=1 -> 128)
+     *
+     * Single-precision (TODO)
+     * ----------------
+     * Assume cmode = 1111 and op = 0
+     *
+     * imm32 = imm8<7>:NOT(imm8<6>):Replicate(imm8<6>,5):imm8<5:0>:Zeros(19);
+     *         a:~b:bbbbb:cdefgh:0000000000000000000
+     *
+     * imm64 = Replicate(imm32, 2);
+     *       = a:~b:bbbbb:cdefgh:0000000000000000000 a:~b:bbbbb:cdefgh:0000000000000000000
+     *
+     * datasize = if Q == '1' then 128 else 64;
+     * imm = Replicate(imm64, datasize DIV 64);
+     *     = imm64       (Q=0)
+     *     = imm64:imm64 (Q=1)
+     */
+    union {
+        __fp16 f;
+        ushort i;
+    } fpv;
+
+    int abc = extract_uint(enc, 16, 3);
+    int defgh = extract_uint(enc, 5, 5);
+
+    uint a = (abc & 0x4);
+    uint b = (abc & 0x2);
+    uint not_b = b == 0 ? 1 : 0;
+
+    uint bb;
+    bb = b == 0 ? 0 : 0x3;
+
+    uint cdefgh = ((abc & 0x1) << 5) | (defgh & 0x1f);
+
+    ushort imm16 = (a << 13) | (not_b << 14) | (bb << 12) | (cdefgh << 6);
+
+    fpv.i = imm16;
+    *opnd = opnd_create_immed_float(fpv.f);
+
+    return true;
+}
+
+static inline bool
+encode_opnd_fpimm8(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
+{
+    /* Based on the IEEE 754-2008 standard but with Arm-specific details that
+     * are left open by the standard. See Arm Architecture Reference Manual.
+     *
+     * Half-precision example
+     *   __   ________
+     * S/exp\/fraction\
+     *  _
+     * abbbcdefgh000000
+     * 0011110000000000 = 1.0
+     *    _
+     *   abbb cdef gh00 0000
+     * 0x8    0    0    0     a
+     * 0x1    0    0    0     b
+     * 0x0    8    0    0     c
+     * 0x0    7    c    0     defgh
+     */
+    union {
+        __fp16 f;
+        ushort i;
+    } fpv;
+
+    if (opnd_is_immed_float(opnd)) {
```

Comment:
suggest: invert and have `if (!opnd_is_immed_float(opnd)) return false;` so the rest of the function is not indented and thus easier to read

Reply:
Good point, done.